src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,428 @@
+    is the list type and the second element is the inner type.
+
+    For dictionaries, a list with three elements is returned, where the first
+    element is the dict type, the second element is the key type, and the
+    third element is the value type.
+
+    For tuples, a list where the fist element is the tuple type and the rest
+    describes the inner types.
+
+    For Union types, a list with the first element being the Union type and
+    the rest describing the inner types.
+
+    Note that for nested types that lack inner type, e.g. list instead of
+    list[str], the inner type is set to inspect._empty. This edge case is
+    handled by _annotation_parse_to_json_schema.
+
+    Examples:
+        str -> [str]
+        Optional[str] -> (str)
+        str | None -> (str)
+
+        list[str] -> [list, [str]]
+        Optional[list[str]] -> (list, [str])
+
+        dict[str, int] -> [dict, [str], [int]]
+        Optional[dict[str, int]] -> (dict, [str], [int])
+
+        list[list[str]] -> [list, [list, str]]
+        list[Optional[list[str]]] -> [list, (list, [str])]
+
+        dict[str, Optional[int]] -> [dict, [str], (int)]
+
+        Union[str, int] -> [Union, [str], [int]]
+
+        tuple[str, int, list[str]] -> [tuple, [str], [int], [list, str]]
+        tuple[Optional[str], int, Optional[list[str]]] -> (str, [int], (list, str))
+
+        list -> [list]
+    """
+    origin = typing.get_origin(annotation)
+    if origin is None:
+        # Either not a nested type or no type hint
+        # inspect._empty is used for parameters without type hints
+        if annotation not in (str, int, float, bool, inspect._empty, dict, list, tuple):
+            raise ValueError(f"Unsupported type hint: {annotation}")
+        return [annotation]
+    if origin is list:
+        inner_type = _parse_annotation(typing.get_args(annotation)[0])
+        return [origin, inner_type]
+    if origin is dict:
+        key_type = _parse_annotation(typing.get_args(annotation)[0])
+        value_type = _parse_annotation(typing.get_args(annotation)[1])
+        return [origin, key_type, value_type]
+    if origin is tuple:
+        return [
+            origin,
+            *[_parse_annotation(arg) for arg in typing.get_args(annotation)],
+        ]
+    if origin is typing.Union:
+        sub_types = typing.get_args(annotation)
+        if sub_types[-1] is type(None):
+            # Union is an Optional type
+            if len(sub_types) == 2:
+                return tuple(_parse_annotation(sub_types[0]))
+            return (
+                origin,
+                *[_parse_annotation(sub_type) for sub_type in sub_types[:-1]],
+            )
+        # Union type
+        return [
+            origin,
+            *[_parse_annotation(sub_type) for sub_type in sub_types],
+        ]
+
+    raise ValueError(f"Unsupported origin: {origin}")
+
+
+def _annotation_parse_to_json_schema(arg: Union[list, tuple]) -> Mapping[str, Union[str, Mapping, Sequence]]:
```

Comment:
I'm very concerned about how brittle this annotation parsing magic is. I suspect there will be lots of edge cases not covered. And what happens if people are using custom objects as types, or not type annotations at all? 

Want to get others thoughts cc @harry-humanloop

Reply:
As mentioned in the comment above, no type annotations are correctly handled.

Custom types are forbidden:
https://github.com/humanloop/humanloop-python/pull/24/files#diff-042309019eab3a15356f047fe2983e50de42ea2e1b2cf9f0c21a7a87f617c4c8R358

Reply:
Ok we will need to communicate in docs to users that their functions can only use native types and see what implication is

Reply:
this does look quite likely to be brittle; hard to follow carefully - could do with more detailed comments in the code itself.

i don't think we should forbid custom types.
are users allowed to specify their own tool kernel and then use custom types?

Reply:
Don't think we should block on this, but at minimum a user should be able to bypass to define their own tool json_schema directly within a decorator arg incase our inference fails due to custom types or whatever.

Lets add to linear so we can resolved this comment @andreibratu

Reply:
https://linear.app/humanloop/issue/ENG-1378/follow-up-on-tool-signature-parsing